File: Leetcode/dynamic_programming/918_MaximumSumCircularSubarray/Solution.java
class Solution {
    public int maxSubarraySumCircular(int[] A) {
        int total = 0, maxSoFar = A[0], maxEndingHere = 0,
            minSoFar = A[0], minEndingHere = 0;
        for(int num: A){
            total += num;
            maxEndingHere = Math.max(num, maxEndingHere+num);
            maxSoFar = Math.max(maxSoFar, maxEndingHere);
            minEndingHere = Math.min(num, minEndingHere+num);
            minSoFar = Math.min(minSoFar, minEndingHere);
        }
        return (maxSoFar<0)? maxSoFar: Math.max(maxSoFar, total-minSoFar);
    }
}

File: Leetcode/dynamic_programming/918_MaximumSumCircularSubarray/Solution.py
class Solution:
    def maxSubarraySumCircular(self, arr: List[int]) -> int:
        total, n = sum(arr), len(arr)
        if(min(arr)>=0): return total
        if(max(arr)<0):  return max(arr)
        max_so_far = max_ending_here = min_so_far = min_ending_here = arr[0]
        for i in range(1, n):
            max_ending_here = max(arr[i], max_ending_here+arr[i])
            max_so_far = max(max_so_far, max_ending_here)
            min_ending_here = min(arr[i], min_ending_here+arr[i])
            min_so_far = min(min_so_far, min_ending_here)
        return max(max_so_far, total-min_so_far)
